src/keywrap/pkcs11/mod.rs

```diff
@@ -0,0 +1,150 @@
+use crate::keywrap::KeyWrapper;
+use crate::config::EncryptConfig;
+use crate::config::DecryptConfig;
+use std::io;
+use std::collections::HashMap;
+use pkcs11_uri::{Pkcs11Uri, PathAttributes, QueryAttributes};
+use crate::utils::{parse_pkcs11_key_file, encrypt_multiple};
+
+struct Pkcs11 {
+}
+
+// Pkcs11KeyFileObject is a representation of the Pkcs11KeyFile with the pkcs11
+// URI as an object
+pub struct Pkcs11KeyFileObject {
+    pub uri: Pkcs11Uri,
+}
+
+
+// Pkcs11Config describes the layout of a pkcs11 config file
+// The file has the following yaml format:
+// module_directories:
+// - /usr/lib64/pkcs11/
+// allowed_module_paths
+// - /usr/lib64/pkcs11/libsofthsm2.so
+struct Pkcs11Config {
+    module_directories: Vec<String>,
+    allowed_module_paths: Vec<String>,
+}
+
+impl KeyWrapper for Pkcs11 {
+    
+    // wrap_keys wraps the session key for recpients and encrypts the opts_data,
+    // which describe the symmetric key used for encrypting the layer
+    fn wrap_keys(&self,
+                 ec: &EncryptConfig,
+                 opts_data: &Vec<u8>)
+                 -> Result<Vec<u8>, io::Error> {
+        let mut x: Vec<Vec<u8>> = Vec::new();
+        let ps: &Vec<Vec<u8>> = &ec.param["pkcs11-pubkeys"];
+        for p in ps {
+            x.push(p.to_vec());
+        }
+        for y in &ec.param["pkcs11-yamsl"] {
+            x.push(y.to_vec());
+        }
+        let pkcs11_recipients: Vec<Pkcs11KeyFileObject>
+          = add_pub_keys(&(ec.decrypt_config.as_ref().unwrap()), &x).unwrap();
+
+        if pkcs11_recipients.len() == 0 {
```

Comment:
vectors have an is_empty method that some people prefer over this kind of check.